### PR TITLE
Improve the `DiscreteFactorFromErrors`

### PR DESCRIPTION
**Pull request by varunagrawal**
Added a new constructor to `AlgebraicDecisionTree` so we can reduce some boilerplate code.

Improved the docstring for `DiscreteFactorFromErrors`.

### CODE REVIEW

File: gtsam/discrete/AlgebraicDecisionTree.h [hunk 1]
```diff
@@ -182,6 +182,21 @@ namespace gtsam {
...
+     *
+     * @tparam X The type of the leaf of the original DecisionTree
+     * @tparam Func Type signature of functional `f`.
+     * @param other The original DecisionTree from which the
+     * AlgbraicDecisionTree is constructed.
+     * @param f Functional used to operate on
+     * the leaves of the input DecisionTree.
+     */
+    template <typename X, typename Func>
+    AlgebraicDecisionTree(const DecisionTree<L, X>& other, Func f)
```
Comment (dellaert) [inline gtsam/discrete/AlgebraicDecisionTree.h:197]: An ADT stores doubles, so ideally Func should be enforced to return a double. We can merge as is but any signature that does not will yield strange compilation errors.

**Review by dellaert — APPROVED** [review]
Small nit on Func, but we can merge as is.

**Comment by dellaert** [comment]
Please merge after merging no_hiding so this targets develop

File: gtsam/discrete/AlgebraicDecisionTree.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/discrete/AlgebraicDecisionTree.h:197]: Hmm I am not sure I know how to enforce that. :( I'll check later.

**Pull request merged**